src/google/google-calendars/googlecalendarsyncadaptor.cpp

```diff
@@ -1737,43 +1821,281 @@ void GoogleCalendarSyncAdaptor::upsyncFinishedHandler()
+                    event->startUpdates();
+                    event->recurrence()->addExDateTime(recurrenceId);
+                    event->endUpdates();
+                } break;
+                case GoogleCalendarSyncAdaptor::Modify: {
+                    SOCIALD_LOG_DEBUG("Event modified remotely:" << eventId);
+                    KCalCore::Event::Ptr event = allLocalEventsMap.value(eventId);
+                    bool changed = false; // modification, not insert, so initially changed = "false".
+                    jsonToKCal(eventData, event, m_serverCalendarIdToDefaultReminderTimes[accountId].value(calendarId), m_icalFormat, &changed);
+                } break;
+                case GoogleCalendarSyncAdaptor::Insert: {
+                    // add a new local event for the remote addition.
+                    KCalCore::Event::Ptr event;
+                    if (recurrenceId.isValid()) {
+                        // this is a persistent occurrence for an already-existing series.
+                        SOCIALD_LOG_DEBUG("Persistent occurrence added remotely:" << eventId);
+                        KCalCore::Event::Ptr parentEvent = allLocalEventsMap.value(parentId);
+                        if (parentEvent.isNull()) {
+                            // it might have been newly added in this sync cycle.  Look for it from the calendar.
+                            QString parentEventUid = m_recurringEventIdToKCalUid.value(accountId).value(parentId);
+                            parentEvent = parentEventUid.isEmpty() ? parentEvent : m_calendar->event(parentEventUid, KDateTime());
+                            if (parentEvent.isNull()) {
+                                SOCIALD_LOG_ERROR("Cannot find parent event:" << parentId << "for persistent occurrence:" << eventId);
+                                m_syncSucceeded[accountId] = false;
+                                continue; // we don't return, but instead attempt to finish other event modifications
+                            }
+                        }
+
+                        // dissociate the persistent occurrence
+                        event = m_calendar->dissociateSingleOccurrence(parentEvent, recurrenceId, recurrenceId.timeSpec()).staticCast<KCalCore::Event>();
+                        if (event.isNull()) {
+                            SOCIALD_LOG_ERROR("Could not dissociate occurrence from recurring event:" << parentId << recurrenceId.toString());
+                            m_syncSucceeded[accountId] = false;
+                            continue; // we don't return, but instead attempt to finish other event modifications
+                        }
+                    } else {
+                        // this is a new event in its own right.
+                        SOCIALD_LOG_DEBUG("Event added remotely:" << eventId);
+                        event = KCalCore::Event::Ptr(new KCalCore::Event);
```

Comment:
Suppose there are two or more devices running this code, event is created and upsynced from one, and then gets downsynced to rest. It would be then new event on those other devices and as mkcal uid is stored on google side, it would need to be used here as well?